### PR TITLE
Replaced ptr::reset( new Class) with make_shared or unique

### PR DESCRIPTION
**Pull request by pvanlaar**
### Description

As described in #2757, cleaned up the `SmartPointer::reset(new ...)`calls. They were replaced with std::make_shared or std::make_unique. Not using new or delete is a C++ best practice ([C++ Core guidelines](http://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Rr-newdelete)). 

### Scope
To keep the PR's relatively concise, I only changed the Ptr::reset( occurrences that could be changed within a few lines of code. There are for example other Ptr::reset() calls that take the result of factory functions producing raw pointers. Cleaning that up would requiring changing the functions and I considered them out of the scope of this PR. If someone has an opinion about the scope, please let me know.

Edit: Removed catkin tools issue, because the CI build for ubuntu passes.

### Checklist
- [x] **Required by CI**: Code is auto formatted using [clang-format](http://moveit.ros.org/documentation/contributing/code)
~~- [] Extend the tutorials / documentation [reference](http://moveit.ros.org/documentation/contributing/)~~ No functional changes
~~- [] Document API changes relevant to the user in the [MIGRATION.md](https://github.com/ros-planning/moveit/blob/master/MIGRATION.md) notes~~ No API changes
~~- [ ] Create tests, which fail without this PR [reference](https://ros-planning.github.io/moveit_tutorials/doc/tests/tests_tutorial.html)~~ No functional changes

### CODE REVIEW

File: moveit_planners/ompl/ompl_interface/src/ompl_planner_manager.cpp
Comment (v4hn) [inline moveit_planners/ompl/ompl_interface/src/ompl_planner_manager.cpp:299]: Please do not remove this comment. If anything, we should remove the whole commented out blocks instead and that would belong into a separate PR.

File: moveit_planners/pilz_industrial_motion_planner/CMakeLists.txt
Comment (v4hn) [inline moveit_planners/pilz_industrial_motion_planner/CMakeLists.txt:5]: This is unrelated and should not be in the same PR/commit. Otherwise it's a good change though.

File: moveit_planners/sbpl/ros/sbpl_interface_ros/src/sbpl_meta_plugin.cpp
Comment (v4hn) [inline moveit_planners/sbpl/ros/sbpl_interface_ros/src/sbpl_meta_plugin.cpp:53]: Why would you remove that line? Same as with the one in OMPL we should cleanup the context for this in a different PR if we remove commented-out code.

File: moveit_ros/perception/mesh_filter/include/moveit/mesh_filter/depth_self_filter_nodelet.h
Comment (v4hn) [inline moveit_ros/perception/mesh_filter/include/moveit/mesh_filter/depth_self_filter_nodelet.h:111]: This moves from `boost` to `std`. Our own updaters do not set this from any API that relies on the boost pointers, so I'm ok with the change in `master`.

File: moveit_ros/planning/rdf_loader/src/rdf_loader.cpp
Comment (v4hn) [inline moveit_ros/planning/rdf_loader/src/rdf_loader.cpp:106]: yay, that one actually fixes potential memory leaks from what I can see :bouquet:

File: moveit_ros/planning_interface/move_group_interface/src/move_group_interface.cpp
Comment (v4hn) [inline moveit_ros/planning_interface/move_group_interface/src/move_group_interface.cpp:1336]: These changes in various places are unrelated, but might be required by clang-tidy?
I don't have a strong opinion on whether it's ok to have them in the same PR though.

**Review by v4hn — APPROVED** [review]
Thank you! I should create such issues more often if there are actually users out there who are willing to contribute such nice cleanups! :1st_place_medal: 

In general I approve. Please reply/address my inline comments though.

I will leave the second review to @tylerjw as he pushed for these changes most in the past :sunglasses:

File: moveit_core/collision_distance_field/src/collision_env_distance_field.cpp
Comment (tylerjw) [inline moveit_core/collision_distance_field/src/collision_env_distance_field.cpp:1091]: Small nit can you change vector operator[] to at()?  Feel free to ignore this and we can do it in a separate change.

**Review by tylerjw — APPROVED** [review]
Generally approve, fix the small things @v4hn asked for and then let's merge this.

File: moveit_planners/pilz_industrial_motion_planner/CMakeLists.txt
Comment (pvanlaar) [reply moveit_planners/pilz_industrial_motion_planner/CMakeLists.txt:5]: This is necessary to use std::make_unique as that function was added later (c++ 14) to the standard than std::make_shared (c++ 11) for some reason. Do you agree that this can be changed here?

File: moveit_planners/ompl/ompl_interface/src/ompl_planner_manager.cpp
Comment (pvanlaar) [reply moveit_planners/ompl/ompl_interface/src/ompl_planner_manager.cpp:299]: Here I react to all comments on the removal of commented out code. I thought to remove these commented lines as they are dead and not subject to CI checks. Commented out code is an [official code smel](https://moderatemisbehaviour.github.io/clean-code-smells-and-heuristics/comments/c5-commented-out-code.html). When I encountered these comments I simply deleted them, as I thought that searching for them later is difficult and cumbersome. I will change all comments back and add them to a separate PR.

File: moveit_ros/perception/mesh_filter/include/moveit/mesh_filter/depth_self_filter_nodelet.h
Comment (pvanlaar) [reply moveit_ros/perception/mesh_filter/include/moveit/mesh_filter/depth_self_filter_nodelet.h:111]: Thank you for clarifying the rationale behind allowing the change. Could you elaborate a bit more, because I don't fully understand it. 

- What are the updaters? Are they clean up scripts? 
- Do you mean with the second half, that these pointers are not exposed publically, so they can be changed with this PR? Otherwise, it has to be done with a deprecation?
- Do you mean that this change should only be in `master` not in `noetic-devel`?

File: moveit_ros/planning_interface/move_group_interface/src/move_group_interface.cpp
Comment (pvanlaar) [reply moveit_ros/planning_interface/move_group_interface/src/move_group_interface.cpp:1336]: Apologies, These changes were done by running clang-format. They were required to be allowed to merge. I don't know how it possible that running clang-format on files I did not alter, changes the file in some places.

File: moveit_core/collision_distance_field/src/collision_env_distance_field.cpp
Comment (pvanlaar) [reply moveit_core/collision_distance_field/src/collision_env_distance_field.cpp:1091]: Good catch. I will change it.

**Comment by pvanlaar** [comment]
Thank you both for the quick reviews and the nice words. One question about the failing Windows build. Do we have to wait on that fix to be allowed to merge?

**Comment by rhaschke** [comment]
The windows build should be fixed. Please, just push your fixups and CI will be re-triggered.

**Comment by pvanlaar** [comment]
Great! Thanks for the quick reply.

File: moveit_planners/ompl/ompl_interface/src/ompl_planner_manager.cpp
Comment (v4hn) [reply moveit_planners/ompl/ompl_interface/src/ompl_planner_manager.cpp]: > Commented out code is an official code smell

I agree! However, in these parts it's actually worse than commented out code.
Some of these commented out functions would actually be useful from my perspective but do not apply anymore because of API changes over time.
In other cases, also in this particular one, the actual functionality is commented out, but not the whole related structure.
The DynamicReconfigure option still exists!

File: moveit_planners/pilz_industrial_motion_planner/CMakeLists.txt
Comment (v4hn) [reply moveit_planners/pilz_industrial_motion_planner/CMakeLists.txt:5]: I see. Just leave it in then. But I would appreciate if you could update the comment above the line too. ;)

In general it's a good change because our whole code base requires C++14 and the pilz planner was added later but not adapted.

File: moveit_ros/perception/mesh_filter/include/moveit/mesh_filter/depth_self_filter_nodelet.h
Comment (v4hn) [reply moveit_ros/perception/mesh_filter/include/moveit/mesh_filter/depth_self_filter_nodelet.h:111]: - We do not differentiate between `master` and `noetic-devel`.
- Ignore the parts about updaters, I thought this change was in a different module when I wrote the comment.
- These are private members of an implementation class, so there is no harm in changing them.

**Review by jliukkonen — APPROVED** [review]
Cool, thanks for cleaning up the codebase!

**Comment by pvanlaar** [comment]
With the latest rebase that includes the latest commits on `noetic-devel` branch, I hope to address the Windows build failure. The PR check clang-tidy did also fail. I ran clang-tidy locally as described in the [docs](https://moveit.ros.org/documentation/contributing/code/), but it fixed many things that were not altered in this PR. I included these fixes in a separate commit. Does anyone know why this happens? I had a similar experience with clang-format.

**Comment by rhaschke** [comment]
> I ran clang-tidy locally as described in the [docs](https://moveit.ros.org/documentation/contributing/code/), but it fixed many things that were not altered in this PR. I included these fixes in a separate commit. Does anyone know why this happens?

Do you use the same versions of clang-tidy / clang-format as in CI?

**Comment by pvanlaar** [comment]
I ran version 10.0.0 of both clang-format and clang-tidy locally, which I understood to be the same versions as the CI build. Correct me if I am wrong.  Exactly the same commands as the [Style guidelines](https://moveit.ros.org/documentation/contributing/code/) were used.

**Comment by jliukkonen** [comment]
>C:\Users\runneradmin\micromamba\envs\test\Library\include\boost/bind.hpp(41): note: The practice of declaring the Bind placeholders (_1, _2, ...) in the global namespace is deprecated. Please use <boost/bind/bind.hpp> + using namespace boost::placeholders, or define BOOST_BIND_GLOBAL_PLACEHOLDERS to retain the current behavior.

While I can't tell you why the Windows build is the only one giving that error, the solution seems to be simple. There's some discussion here: [https://stackoverflow.com/a/65813178](https://stackoverflow.com/a/65813178)

Basically you would do this:
```
boost::bind(&MoveGroupSequenceAction::executeSequenceCallback, this, boost::placeholders::_1)
```
instead of this:
```
boost::bind(&MoveGroupSequenceAction::executeSequenceCallback, this, _1)
```
And then you would also change the include to point to `#include <boost/bind/bind.hpp>` instead of `#include <boost/bind.hpp>`. That fix doesn't seem unreasonable to apply to the codebase although it will end up being probably 50+ changes.

**Comment by rhaschke** [comment]
The boost placeholders are just warnings (probably because the Windows build uses a more recent boost version).
The actual error is this one:
`C:\Program Files (x86)\Microsoft Visual Studio\2019\Enterprise\VC\Tools\MSVC\14.16.27023\include\type_traits(1271): error C2338: You've instantiated std::aligned_storage<Len, Align> with an extended alignment (in other words, Align > alignof(max_align_t)). Before VS 2017 15.8, the member type would non-conformingly have an alignment of only alignof(max_align_t). VS 2017 15.8 was fixed to handle this correctly, but the fix inherently changes layout and breaks binary compatibility (*only* for uses of aligned_storage with extended alignments). Please define either (1) _ENABLE_EXTENDED_ALIGNED_STORAGE to acknowledge that you understand this message and that you actually want a type with an extended alignment, or (2) _DISABLE_EXTENDED_ALIGNED_STORAGE to silence this message and get the old non-conformant behavior.`

**Comment by rhaschke** [comment]
Weird, that there are still clang-tidy issues...

**Comment by jliukkonen** [comment]
> The boost placeholders are just warnings (probably because the Windows build uses a more recent boost version).
> The actual error is this one:
> `C:\Program Files (x86)\Microsoft Visual Studio\2019\Enterprise\VC\Tools\MSVC\14.16.27023\include\type_traits(1271): error C2338: You've instantiated std::aligned_storage<Len, Align> with an extended alignment (in other words, Align > alignof(max_align_t)). Before VS 2017 15.8, the member type would non-conformingly have an alignment of only alignof(max_align_t). VS 2017 15.8 was fixed to handle this correctly, but the fix inherently changes layout and breaks binary compatibility (*only* for uses of aligned_storage with extended alignments). Please define either (1) _ENABLE_EXTENDED_ALIGNED_STORAGE to acknowledge that you understand this message and that you actually want a type with an extended alignment, or (2) _DISABLE_EXTENDED_ALIGNED_STORAGE to silence this message and get the old non-conformant behavior.`

Ah yes, apologies for my misleading comment @pvanlaar. However, if you do end up fixing the Boost related warnings too, that would be a nice improvement. As for this error, I don't know which one of those solutions would be the correct one.

**Comment by pvanlaar** [comment]
Thank you both for the replies. Don't worry @jliukkonen, I appreciate the help. Issue #2766 would solve those boost::bind warnings, so we won't have to create a separate issue for that. I will have a look at the current merge fails.

**Comment by rhaschke** [comment]
@Tobias-Fischer and/or @wolfv: Could you have a look at https://github.com/ros-planning/moveit/pull/2762#issuecomment-880936888 and suggest a solution?
My gut feeling is to add the flag `_ENABLE_EXTENDED_ALIGNED_STORAGE` via build options in the CI script.

**Comment by wolfv** [comment]
Thanks for the ping. I think the _ENABLE_EXTENDED_ALIGNED_STORAGE is probably what we want, longterm. 
I cannot really judge if this will have implications on the ABI of MoveIt. 
Ideally this flag would be applied somewhere higher level (e.g. catkin) so that all of ROS uses the same compiler option.

I think the flag is probably most cleanly added from CMake, as we're doing here:

https://github.com/RoboStack/ros-noetic/blob/cbd8a2011de90923a154bca7533d0e1578269065/patch/ros-noetic-geometric-shapes.win.patch#L23-L27

**Comment by pvanlaar** [comment]
Thank you for the assistance @rhaschke. Nice solution to do this with the help of clang-tidy.

**Pull request merged**
